libbeat/common/kubernetes/util.go

```diff
@@ -94,54 +98,65 @@ func IsInCluster(kubeconfig string) bool {
 // If host is provided in the config use it directly.
 // If beat is deployed in k8s cluster, use hostname of pod which is pod name to query pod meta for node name.
 // If beat is deployed outside k8s cluster, use machine-id to match against k8s nodes for node name.
-func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (node string) {
+// If node cannot be discovered, return NODE_NAME env var as default value. In case it is not set return error.
+func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (string, error) {
 	if host != "" {
 		log.Infof("kubernetes: Using node %s provided in the config", host)
-		return host
+		return host, nil
+	}
+	nodeNameEnv := os.Getenv("NODE_NAME")
+	var envError, logerror error
+	if nodeNameEnv == "" {
+		envError = errors.New("kubernetes: NODE_NAME environment variable was not set")
 	}
 	ctx := context.TODO()
 	if inCluster {
 		ns, err := InClusterNamespace()
 		if err != nil {
-			log.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
+			log.Error(logerror)
+			return nodeNameEnv, errors.Wrap(envError, logerror.Error())
 		}
-		podName, err := os.Hostname()
+		podName, err := osHostname()
 		if err != nil {
-			log.Errorf("kubernetes: Couldn't get hostname as beat pod name in cluster with error: %+v", err.Error())
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Couldn't get hostname as beat pod name in cluster with error: %+v", err.Error())
+			log.Error(logerror)
+			return nodeNameEnv, errors.Wrap(envError, logerror.Error())
 		}
 		log.Infof("kubernetes: Using pod name %s and namespace %s to discover kubernetes node", podName, ns)
 		pod, err := client.CoreV1().Pods(ns).Get(ctx, podName, metav1.GetOptions{})
 		if err != nil {
-			log.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
+			log.Error(logerror)
+			return nodeNameEnv, errors.Wrap(envError, logerror.Error())
 		}
 		log.Infof("kubernetes: Using node %s discovered by in cluster pod node query", pod.Spec.NodeName)
-		return pod.Spec.NodeName
+		return pod.Spec.NodeName, nil
 	}
 
-	mid := machineID()
+	mid := machineId()
 	if mid == "" {
-		log.Error("kubernetes: Couldn't collect info from any of the files in /etc/machine-id /var/lib/dbus/machine-id")
-		return defaultNode
+		logerror = errors.New("kubernetes: Couldn't collect info from any of the files in /etc/machine-id /var/lib/dbus/machine-id")
+		log.Error(logerror)
+		return nodeNameEnv, errors.Wrap(envError, logerror.Error())
 	}
 
 	nodes, err := client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
```

Comment:
Maybe we can add a field selector here, so we don't need to loop over all the nodes in the cluster, something like this (not tested):
```suggestion
	nodes, err := client.CoreV1().Nodes().List(ctx, metav1.ListOptions{FieldSelector: fmt.Sprintf("status.nodeInfo.machineID=%s", mid)})
```

Reply:
I tried but unfortunately that filed is not supported as a field-selector.  This is the list of supported selectors for node : https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/conversion.go#L60